src/dtype.rs

```diff
@@ -94,12 +106,202 @@ impl PyArrayDescr {
+    ///
+    /// The `shape` is the fixed shape of the sub-array described by this data type,
+    /// and `item_dtype` the data type of the array.
+    ///
+    /// If a field whose dtype object has this attribute is retrieved, then the extra dimensions
+    /// implied by shape are tacked on to the end of the retrieved array.
+    ///
+    /// Equivalent to [`np.dtype.subdtype`](https://numpy.org/doc/stable/reference/generated/numpy.dtype.subdtype.html)
+    pub fn subdtype(&self) -> Option<(&PyArrayDescr, Vec<usize>)> {
+        self.shape()
+            .and_then(|shape| self.base().map(|base| (base, shape)))
+    }
+
+    /// Returns true if the dtype is a sub-array at the top level.
+    ///
+    /// Equivalent to [`np.dtype.hasobject`](https://numpy.org/doc/stable/reference/generated/numpy.dtype.hasobject.html)
+    pub fn has_object(&self) -> bool {
+        self.flags() & NPY_ITEM_HASOBJECT != 0
+    }
+
+    /// Returns true if the dtype is a struct which maintains field alignment.
+    ///
+    /// This flag is sticky, so when combining multiple structs together, it is preserved
+    /// and produces new dtypes which are also aligned.
+    ///
+    /// Equivalent to [`np.dtype.isalignedstruct`](https://numpy.org/doc/stable/reference/generated/numpy.dtype.isalignedstruct.html)
+    pub fn is_aligned_struct(&self) -> bool {
+        self.flags() & NPY_ALIGNED_STRUCT != 0
+    }
+
+    /// Returns true if the data type is a sub-array.
+    pub fn has_subarray(&self) -> bool {
+        // equivalent to PyDataType_HASSUBARRAY(self)
+        unsafe { !(*self.as_dtype_ptr()).subarray.is_null() }
+    }
+
+    /// Returns true if the data type is a structured type.
+    pub fn has_fields(&self) -> bool {
+        // equivalent to PyDataType_HASFIELDS(self)
+        unsafe { !(*self.as_dtype_ptr()).names.is_null() }
+    }
+
+    /// Returns true if the data type is unsized
+    pub fn is_unsized(&self) -> bool {
+        // equivalent to PyDataType_ISUNSIZED(self)
+        self.itemsize() == 0 && !self.has_fields()
+    }
+
+    /// Returns true if data type byteorder is native, or `None` if not applicable.
+    pub fn is_native_byteorder(&self) -> Option<bool> {
+        // based on PyArray_ISNBO(self->byteorder)
+        match self.byteorder() {
+            b'=' => Some(true),
+            b'|' => None,
+            byteorder if byteorder == NPY_BYTEORDER_CHAR::NPY_NATBYTE as u8 => Some(true),
+            _ => Some(false),
+        }
+    }
+
+    /// Returns an ordered list of field names, or `None` if there are no fields.
+    ///
+    /// The names are ordered according to increasing byte offset.
+    ///
+    /// Equivalent to [`np.dtype.names`](https://numpy.org/doc/stable/reference/generated/numpy.dtype.names.html).
+    pub fn names(&self) -> Option<Vec<&str>> {
+        if !self.has_fields() {
+            return None;
+        }
+        let names = unsafe { PyTuple::from_borrowed_ptr(self.py(), (*self.as_dtype_ptr()).names) };
```

Comment:
Maybe my lack of `pyo3` knowledge, looks like an overkill to incref here for just one line of code? Is it possible to do it without `from_borrowed_ptr`, or is this the intended way to do it?

Reply:
Are you sure there is reference increment involved here, c.f. https://docs.rs/pyo3/0.15.1/src/pyo3/conversion.rs.html#543-548? I think that happens only when creating e.g. `Py<T>`?

Reply:
Ohh, fair enough - yea, I was looking at source code for the wrong impl. It's all ok then.